### PR TITLE
Add "More" block.

### PR DESCRIPTION
**Pull request by mtias**
image

See #983.

### CODE REVIEW

**Comment by jasmussen** [comment]
So this is interesting... the "READ MORE" text is editable.

Was this intentional on your part? I'm leaning towards this text _not_ being editable, and letting themes decide the text that goes there. But it's a very intriguing idea.

**Comment by jasmussen** [comment]
I dig the help text in the inspector! ⭐️⭐️⭐️⭐️⭐️

**Comment by mtias** [comment]
@jasmussen it is editable on purpose. WordPress handles it like this:
`<!--more Your custom more text... -->`

**Comment by jasmussen** [comment]
Interesting! I did not know that. 

Then I'd like to redesign it a bit. I'll assign myself this branch and do it in code, if that's okay with you.

**Comment by mtias** [comment]
Sure thing. My plan was that the default text is not a placeholder but real text. Not sure it's trivial to get the text a theme has specified, so we may need to present it more abstractly.

**Comment by jasmussen** [comment]
> Not sure it's trivial to get the text a theme has specified, so we may need to present it more abstractly.

I think this is key to figure out if we actually want to make it editable. 

What exactly does the "Your custom more text..." do in `<!--more Your custom more text... -->`? Does it override the Read More text provided by the theme? And if yes, is this a feature that we think will stick around for a bit? It sort of has that "abandoned" smell on it.

In case it's a feature we think will stick around, and it already works today, then I totally love the idea that not only can you customize it but do it on a per-post basis.

The key thing to think about for the design (you touch on it with the "get the text a theme has specified"), is making the configurability discoverable. If we can't actually read the text that the theme has provided, then we should either go super abstract and perhaps not even show the "Read More" text on the block, but something else entirely. Or maybe we do show that, but only let you customize the read more text in the inspector. Something I think I'll need to sleep on.

**Comment by mtias** [comment]
Yes, it works, but is quite the hidden feature:

image

I actually love that we could surface this way better with blocks, and that it's much more obvious how it works _when you can edit the text directly_.

**Comment by dmsnell** [comment]
I'm proposing that we leave this as its own block for now and iterate if necessary later. Essentially `<!--something-->` becomes shorthand for a block comment header. We will need to whitelist the available ones so we don't accidentally capture HTML comments.

```js
// <!-- more -->

{
  blockName: 'wp:core/more'
}

// <!-- more read on! -->

{
  blockName: 'wp:core/more',
  attrs: {
    customText: 'read on!'
  }
}
```

**Comment by swissspidy** [comment]
Can we somehow enforce that this block only can be used once in a post? Something for a separate PR, but just thinking out loud here.

**Comment by mtias** [comment]
@swissspidy ah, yes, definitely.

**Comment by jasmussen** [comment]
I pushed a little polish to this one. Now the text area doesn't span the full width. I also tweaked the description a bit. 

I can't think of a better way to design this block so input is editable. It's not as discoverable as I'd like it to be, but I think we should get this in and iterate later.

Should we add a placeholder? Right now if you delete the text, it's just a dashed line.

**Comment by swissspidy** [comment]
> Should we add a placeholder? Right now if you delete the text, it's just a dashed line.

I think so. "Read more" makes sense. Looking at `get_the_content()`, `(more…)` seems to be the default for the link. However, that doesn't really work well as a placeholder.

File: blocks/library/more/index.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+					value={ text || __( 'Read more' ) }
+					focus={ focus }
+					onFocus={ setFocus }
+					onChange={ ( value ) => setAttributes( { text: value } ) }
+					inline
+					formattingControls={ [] }
+				/>
+				{ focus &&
+					<InspectorControls key="inspector">
+						<p className="editor-block-inspector__description">"More" allows you to break your post into a part shown on index pages, and the subsequent after clicking a "Read More" link.</p>
```
Comment (swissspidy) [inline blocks/library/more/index.js]: This could use the new `<BlockDescription>` element

**Comment by westonruter** [comment]
> Can we somehow enforce that this block only can be used once in a post? Something for a separate PR, but just thinking out loud here.

See #1661.

**Comment by youknowriad** [comment]
I pushed some changes here (essentially the `useOnce` and some polish). I think this is ready to go, can we have some :eyes: here?

**Comment by jasmussen** [comment]
I like it! I think it's good to go. 

There's some visual polish I'd like to do, but that's best done after https://github.com/WordPress/gutenberg/pull/2001 gets merged as the dimensions of the smallest block are tweaked in that. It also feels weird that you can press enter inside the editable read more text, but I think that's also worth trying as is, and then discussing separately.

👍 👍

**Comment by youknowriad** [comment]
>  It also feels weird that you can press enter inside the editable read more text, but I think that's also worth trying as is, and then discussing separately.

Yes, the default behaviour for "inline" editables now is creating `<br />` when we hit enter if splitting the block is not allowed (like the text block), Maybe we should do nothing in this case but this is outside of the scope of this PR (we could also make it opt-in or opt-out). cc @iseulde

**Comment by mtias** [comment]
One other thing that is weird is that the default text is replaced by the theme if you don't edit it. Maybe switching to "read more" not being editable directly, and adding a text field in the inspector, and only if you add custom text to the field in the inspector, then we update the text in the visual block.

But that can be done afterwards, let's ship this @youknowriad.

File: blocks/library/more/index.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+					</BlockDescription>
+					<ToggleControl
+						label={ __( 'Hide the teaser before the "More" tag' ) }
+						checked={ !! noTeaser }
+						onChange={ toggleNoTeaser }
+					/>
+				</InspectorControls>
+			),
+			<div key="more-tag" className="wp-block-more">
+				<Editable
```
Comment (aduth) [inline blocks/library/more/index.js]: Why is this Editable, and not just a simple `<input type="text" />` ?

**Comment by youknowriad** [comment]
> One other thing that is weird is that the default text is replaced by the theme

I'm thinking we could leverage the editor settings to be introduced here to fix this https://github.com/WordPress/gutenberg/pull/2021. Should be done in a follow-up PR

File: blocks/test/fixtures/core__more.html [hunk 3]
```diff
@@ -0,0 +1 @@
+<!-- wp:core/more {"text": "Continue reading", "noTeaser": true} /-->
```
Comment (nylen) [inline blocks/test/fixtures/core__more.html]: How did you generate this?  Since it won't do anything in WP core - and since it's different from the default serialization - I think it should use the `<!--more-->` format.

File: post-content.js [hunk 4]
```diff
@@ -26,8 +26,6 @@ window._wpGutenbergPost.content = {
 		'<p style="text-align:right;">... like this one, which is right aligned.</p>',
 		'<!-- /wp:core/text -->',
 
-		'<!-- wp:core/more /-->',
-
```
Comment (nylen) [inline post-content.js]: Same comment as above, I think this should use the old-style `<!--more-->` format.

**Comment by nylen** [comment]
This is a pretty weird block from a parsing/serialization standpoint.  Let's be sure to add tests for its different variations:

- `<!-- wp:core/more -->` - this will fail the validation added in #1929, which is why the Travis build is failing.
- No text and without the `noteaser` (simply `<!--more-->`)
- Maybe another invalid variation such as `<!--more-->\nsome text<!--noteaser-->` - how likely is this to happen?

**Comment by mtias** [comment]
I also don't like that we expose `noteaser` in the UI so prominently now (in the inspector). Not something for this PR, and @jasmussen is looking at an advanced panel for the inspector for some of these settings. But we may also consider not showing this one.

**Review by nylen — DISMISSED** [review]
As noted above, needs rebase, test suite fixes, and tests for the different markup variations.

File: blocks/library/more/index.js [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+						<p>{ __( '"More" allows you to break your post into a part shown on index pages, and the subsequent after clicking a "Read More" link.' ) }</p>
+					</BlockDescription>
+					<ToggleControl
+						label={ __( 'Hide the teaser before the "More" tag' ) }
+						checked={ !! noTeaser }
+						onChange={ toggleNoTeaser }
+					/>
+				</InspectorControls>
+			),
+			<div key="more-tag" className="wp-block-more">
```
Comment (aduth) [inline blocks/library/more/index.js]: Should this be `props.className` instead of `"wp-block-more"` ?

File: editor/assets/stylesheets/_z-index.scss [hunk 6]
```diff
@@ -5,6 +5,7 @@
 $z-layers: (
 	'.editor-block-switcher__arrow': 1,
 	'.editor-visual-editor__block:before': -1,
+	'.editor-visual-editor__block .wp-block-more before': -1,
```
Comment (aduth) [inline editor/assets/stylesheets/_z-index.scss]: I guess it doesn't matter much, but other keys leave the `:` (colon) of `:before`

File: blocks/test/fixtures/core__more.html (hunk 3 above)
Comment (youknowriad) [inline blocks/test/fixtures/core__more.html]: Right! this was generating using `GENERATE_MISSING_FIXTURES=y npm run test-unit` maybe we need to change something in the command line

File: post-content.js (hunk 4 above)
Comment (youknowriad) [inline post-content.js]: There's already a `more` above in the post-content, that's why I dropped this.

File: blocks/library/more/index.js (hunk 5 above)
Comment (youknowriad) [inline blocks/library/more/index.js]: no, when we set the `className` to false, there's no className prop

File: blocks/test/fixtures/core__more.html (hunk 3 above)
Comment (youknowriad) [inline blocks/test/fixtures/core__more.html]: Oh I guess, this was added before my changes, I'm updating to the last format

Comment (youknowriad) [inline blocks/test/fixtures/core__more.html]: Edit: The issue was that we were checking that each block as a corresponding test file and this file contains `<!-- wp:{ name }` with a regex, I'm relying on the JSON file now to avoid recreating the parsing mechanism in the test

Comment (youknowriad) [inline blocks/test/fixtures/core__more.html]: I couldn't add the invalid unit test because it breaks the test because it's being parsed to freeform instead of more.

**Comment by youknowriad** [comment]
Ok merging. We can iterate in smaller PRs

File: blocks/library/more/index.js (hunk 5 above)
Comment (aduth) [inline blocks/library/more/index.js]: Why do we set the `className` to false if we clearly want one? Is this related to `save` auto-applying the class name?

Comment (youknowriad) [inline blocks/library/more/index.js]: it's more related to the className input showing in the inspector if the className is not set to false

**Comment by swissspidy** [comment]
> Maybe another invalid variation such as `<!--more-->\nsome text<!--noteaser-->` - how likely is this to happen?

It's not invalid in PHP. WordPress doesn't care where in the content `<!--noteaser-->` is. It doesn't need to follow right after the more tag. How should legacy content be handled in this scenario?

**Comment by iseulde** [comment]
I haven't followed the discussion closely, and maybe I'm just repeating what someone else said, but what if we put the legacy content in the block body, and upgrade old ones when they're between blocks?

```html
<!-- wp:core/more { "text": "some text...", "noTeaser":true } -->
<!--more some text...-->
<!--noteaser-->
<!-- /wp:core/more -->
```

**Comment by aduth** [comment]
@iseulde I'm not really sure. I like your proposal for consistency's sake. Maybe there's an argument for special-casing some "known" types.

**Comment by nylen** [comment]
Another way that would be cleaner: make core recognize the `<!-- wp:core/more /-->` syntax using our server-side parser.  I'm not sure how difficult this would be.

**Comment by mtias** [comment]
I agree with @nylen, let's see if we can consolidate all comments under the same syntax. Doesn't make much sense to me to wrap an old comment in new comments, it just adds noise.
